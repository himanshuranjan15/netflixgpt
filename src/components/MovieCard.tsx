import React from 'react';
import { IMG_CDN_URL } from '../utils/constants';

interface MovieCardProps {
  posterPath: string;
  title: string;
}

const MovieCard: React.FC<MovieCardProps> = ({ posterPath, title }) => {
  if (!posterPath) return null;

  return (
    <div className="w-48 pr-4">
      <img alt={title} src={IMG_CDN_URL + posterPath} />
    </div>
  );
};

export default MovieCard;
