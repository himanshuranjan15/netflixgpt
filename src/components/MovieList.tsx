import React from 'react';
import MovieCard from './MovieCard';

interface Movie {
  id: number;
  poster_path: string;
  title: string;
}

interface MovieListProps {
  title: string;
  movies: Movie[];
}

const MovieList: React.FC<MovieListProps> = ({ title, movies }) => {
  // If movies is not provided or is empty, don't render the list part,
  // but still render the title container for consistency, or return null if title shouldn't show.
  // For this case, let's assume the title should always show, and an empty list means no movie cards.
  if (!movies) { // Handles null or undefined
    return (
      <div className="px-6">
        <h1 className="text-3xl py-4 text-white">{title}</h1>
        <div className="flex overflow-x-scroll">
          <div className="flex">
            {/* No movies to map */}
          </div>
        </div>
      </div>
    );
  }

  return (
    <div className="px-6">
      <h1 className="text-3xl py-4 text-white">{title}</h1>
      <div className="flex overflow-x-scroll">
        <div className="flex">
          {movies.map((movie) => (
            <MovieCard key={movie.id} posterPath={movie.poster_path} title={movie.title} />
          ))}
        </div>
      </div>
    </div>
  );
};

export default MovieList;
