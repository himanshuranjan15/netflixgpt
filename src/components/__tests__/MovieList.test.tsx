import React from 'react';
import { render, screen } from '@testing-library/react';
import MovieList from '../MovieList'; // Adjust path as necessary
import MovieCard from '../MovieCard'; // Import the mocked MovieCard

// Mock MovieCard to simplify MovieList tests and avoid testing MovieCard's internals here
jest.mock('../MovieCard', () => {
  // Mocking the default export of MovieCard
  return jest.fn(({ posterPath, title }) => ( // Ensure mock signature matches usage
    <div data-testid="movie-card">
      <img src={posterPath} alt={title} />
      <p>{title}</p>
    </div>
  ));
});

describe('MovieList Component', () => {
  const mockMovies = [
    { id: 1, poster_path: '/poster1.jpg', title: 'Movie 1' },
    { id: 2, poster_path: '/poster2.jpg', title: 'Movie 2' },
    { id: 3, poster_path: '/poster3.jpg', title: 'Movie 3' },
  ];

  const listTitle = 'Popular Movies';

  test('renders the list title correctly', () => {
    render(<MovieList title={listTitle} movies={mockMovies} />);
    const titleElement = screen.getByText(listTitle);
    expect(titleElement).toBeInTheDocument();
    expect(titleElement.tagName).toBe('H1');
  });

  test('renders a list of MovieCard components', () => {
    render(<MovieList title={listTitle} movies={mockMovies} />);
    const movieCards = screen.getAllByTestId('movie-card');
    expect(movieCards).toHaveLength(mockMovies.length);
  });

  test('passes correct props to MovieCard components', () => {
    render(<MovieList title={listTitle} movies={mockMovies} />);

    const mockMovieCardInstance = MovieCard as jest.Mock;

    expect(mockMovieCardInstance.mock.calls.length).toBe(mockMovies.length);
    mockMovies.forEach((movie, index) => {
      const callArgs = mockMovieCardInstance.mock.calls[index];
      expect(callArgs[0]).toEqual(expect.objectContaining({ // Changed to toEqual for the props object
        posterPath: movie.poster_path,
        title: movie.title,
      }));
      // Optionally, assert the second argument if its state is consistent and important
      // For example, if it should always be undefined:
      // expect(callArgs[1]).toBeUndefined();
    });
  });

  test('renders an empty list correctly if movies array is empty', () => {
    render(<MovieList title={listTitle} movies={[]} />);
    const movieCards = screen.queryAllByTestId('movie-card');
    expect(movieCards).toHaveLength(0);
    // Check that the title still renders
    expect(screen.getByText(listTitle)).toBeInTheDocument();
  });

  test('renders correctly if movies array is undefined', () => {
    render(<MovieList title={listTitle} movies={undefined as any} />);
    const movieCards = screen.queryAllByTestId('movie-card');
    expect(movieCards).toHaveLength(0);
    expect(screen.getByText(listTitle)).toBeInTheDocument();
  });

  test('renders correctly if movies array is null', () => {
    render(<MovieList title={listTitle} movies={null as any} />);
    const movieCards = screen.queryAllByTestId('movie-card');
    expect(movieCards).toHaveLength(0);
    expect(screen.getByText(listTitle)).toBeInTheDocument();
  });

  // Snapshot test (optional, but good for catching UI regressions)
  test('matches snapshot', () => {
    const { asFragment } = render(<MovieList title={listTitle} movies={mockMovies} />);
    expect(asFragment()).toMatchSnapshot();
  });

   test('matches snapshot with empty movies list', () => {
    const { asFragment } = render(<MovieList title={listTitle} movies={[]} />);
    expect(asFragment()).toMatchSnapshot();
  });

});
