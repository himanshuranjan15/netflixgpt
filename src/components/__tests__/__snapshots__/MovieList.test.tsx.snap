// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MovieList Component matches snapshot 1`] = `
<DocumentFragment>
  <div
    class="px-6"
  >
    <h1
      class="text-3xl py-4 text-white"
    >
      Popular Movies
    </h1>
    <div
      class="flex overflow-x-scroll"
    >
      <div
        class="flex"
      >
        <div
          data-testid="movie-card"
        >
          <img
            alt="Movie 1"
            src="/poster1.jpg"
          />
          <p>
            Movie 1
          </p>
        </div>
        <div
          data-testid="movie-card"
        >
          <img
            alt="Movie 2"
            src="/poster2.jpg"
          />
          <p>
            Movie 2
          </p>
        </div>
        <div
          data-testid="movie-card"
        >
          <img
            alt="Movie 3"
            src="/poster3.jpg"
          />
          <p>
            Movie 3
          </p>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`MovieList Component matches snapshot with empty movies list 1`] = `
<DocumentFragment>
  <div
    class="px-6"
  >
    <h1
      class="text-3xl py-4 text-white"
    >
      Popular Movies
    </h1>
    <div
      class="flex overflow-x-scroll"
    >
      <div
        class="flex"
      />
    </div>
  </div>
</DocumentFragment>
`;
